File: Loon-0.5-Alpha/Loon-Neo/src/loon/utils/TArray.java
package loon.utils;

import java.util.Iterator;
import java.util.NoSuchElementException;

import loon.utils.ObjectMap.Keys;
import loon.utils.ObjectMap.Values;

@SuppressWarnings({ "rawtypes", "unchecked" })
public class TArray<T> implements Iterable<T> {

	public T[] items;

	public int size;
	public boolean ordered;

	public TArray() {
		this(true, CollectionUtils.INITIAL_CAPACITY);
	}

	public TArray(int capacity) {
		this(true, capacity);
	}

	public TArray(boolean ordered, int capacity) {
		this.ordered = ordered;
		items = (T[]) new Object[capacity];
	}

	public TArray(TArray<? extends T> array) {
		this(array.ordered, array.size);
		size = array.size;
		System.arraycopy(array.items, 0, items, 0, size);
	}

	public TArray(T[] array) {
		this(true, array, 0, array.length);
	}

	public TArray(boolean ordered, T[] array, int start, int count) {
		this(ordered, count);
		size = count;
		System.arraycopy(array, start, items, 0, size);
	}

	public TArray(SortedList<T> vals) {
		this();
		for (LIterator<T> it = vals.listIterator(); it.hasNext();) {
			add(it.next());
		}
	}
	
	public TArray(Keys<T> vals) {
		this();
		for (T t : vals) {
			add(t);
		}
	}
	
	public TArray(Values<T> vals) {
		this();
		for (T t : vals) {
			add(t);
		}
	}

	public boolean add(T value) {
		T[] items = this.items;
		if (size == items.length) {
			items = resize(MathUtils.max(8, (int) (size * 1.75f)));
		}
		items[size++] = value;
		return true;
	}

	public void addAll(TArray<? extends T> array) {
		addAll(array, 0, array.size);
	}

	public void addAll(TArray<? extends T> array, int start, int count) {
		if (start + count > array.size)
			throw new IllegalArgumentException(
					"start + count must be <= size: " + start + " + " + count
							+ " <= " + array.size);
		addAll((T[]) array.items, start, count);
	}

	public void addAll(T... array) {
		addAll(array, 0, array.length);
	}

	public void addAll(T[] array, int start, int count) {
		T[] items = this.items;
		int sizeNeeded = size + count;
		if (sizeNeeded > items.length)
			items = resize(MathUtils.max(8, (int) (sizeNeeded * 1.75f)));
		System.arraycopy(array, start, items, size, count);
		size += count;
	}

	public T get(int index) {
		if (index >= size)
			throw new IndexOutOfBoundsException("index can't be >= size: "
					+ index + " >= " + size);
		return items[index];
	}

	public void set(int index, T value) {
		if (index >= size)
			throw new IndexOutOfBoundsException("index can't be >= size: "
					+ index + " >= " + size);
		items[index] = value;
	}

	public void insert(int index, T value) {
		if (index > size)
			throw new IndexOutOfBoundsException("index can't be > size: "
					+ index + " > " + size);
		T[] items = this.items;
		if (size == items.length)
			items = resize(MathUtils.max(8, (int) (size * 1.75f)));
		if (ordered)
			System.arraycopy(items, index, items, index + 1, size - index);
		else
			items[size] = items[index];
		size++;
		items[index] = value;
	}

	public void swap(int first, int second) {
		if (first >= size)
			throw new IndexOutOfBoundsException("first can't be >= size: "
					+ first + " >= " + size);
		if (second >= size)
			throw new IndexOutOfBoundsException("second can't be >= size: "
					+ second + " >= " + size);
		T[] items = this.items;
		T firstValue = items[first];
		items[first] = items[second];
		items[second] = firstValue;
	}

	public boolean contains(T value) {
		return contains(value, false);
	}

	public boolean contains(T value, boolean identity) {
		T[] items = this.items;
		int i = size - 1;
		if (identity || value == null) {
			while (i >= 0)
				if (items[i--] == value)
					return true;
		} else {
			while (i >= 0)
				if (value.equals(items[i--]))
					return true;
		}
		return false;
	}

	public int indexOf(T value) {
		return indexOf(value, false);
	}

	public int indexOf(T value, boolean identity) {
		T[] items = this.items;
		if (identity || value == null) {
			for (int i = 0, n = size; i < n; i++)
				if (items[i] == value)
					return i;
		} else {
			for (int i = 0, n = size; i < n; i++)
				if (value.equals(items[i]))
					return i;
		}
		return -1;
	}

	public int lastIndexOf(T value, boolean identity) {
		T[] items = this.items;
		if (identity || value == null) {
			for (int i = size - 1; i >= 0; i--)
				if (items[i] == value)
					return i;
		} else {
			for (int i = size - 1; i >= 0; i--)
				if (value.equals(items[i]))
					return i;
		}
		return -1;
	}

	public boolean removeValue(T value, boolean identity) {
		T[] items = this.items;
		if (identity || value == null) {
			for (int i = 0, n = size; i < n; i++) {
				if (items[i] == value) {
					removeIndex(i);
					return true;
				}
			}
		} else {
			for (int i = 0, n = size; i < n; i++) {
				if (value.equals(items[i])) {
					removeIndex(i);
					return true;
				}
			}
		}
		return false;
	}

	public T removeIndex(int index) {
		if (index >= size)
			throw new IndexOutOfBoundsException("index can't be >= size: "
					+ index + " >= " + size);
		T[] items = this.items;
		T value = (T) items[index];
		size--;
		if (ordered)
			System.arraycopy(items, index + 1, items, index, size - index);
		else
			items[index] = items[size];
		items[size] = null;
		return value;
	}

	public void removeRange(int start, int end) {
		if (end >= size)
			throw new IndexOutOfBoundsException("end can't be >= size: " + end
					+ " >= " + size);
		if (start > end)
			throw new IndexOutOfBoundsException("start can't be > end: "
					+ start + " > " + end);
		T[] items = this.items;
		int count = end - start + 1;
		if (ordered)
			System.arraycopy(items, start + count, items, start, size
					- (start + count));
		else {
			int lastIndex = this.size - 1;
			for (int i = 0; i < count; i++)
				items[start + i] = items[lastIndex - i];
		}
		size -= count;
	}

	public boolean remove(T value) {
		return remove(value, false);
	}

	public boolean remove(T value, boolean identity) {
		Object[] items = this.items;
		if (identity || value == null) {
			for (int i = 0; i < size; i++) {
				if (items[i] == value) {
					removeIndex(i);
					return true;
				}
			}
		} else {
			for (int i = 0; i < size; i++) {
				if (value.equals(items[i])) {
					removeIndex(i);
					return true;
				}
			}
		}
		return false;
	}

	public boolean removeAll(TArray<? extends T> array) {
		return removeAll(array, false);
	}

	public boolean removeAll(TArray<? extends T> array, boolean identity) {
		if (array.size == 0) {
			return true;
		}
		int size = this.size;
		int startSize = size;
		T[] items = this.items;
		if (identity) {
			for (int i = 0, n = array.size; i < n; i++) {
				T item = array.get(i);
				for (int ii = 0; ii < size; ii++) {
					if (item == items[ii]) {
						removeIndex(ii);
						size--;
						break;
					}
				}
			}
		} else {
			for (int i = 0, n = array.size; i < n; i++) {
				T item = array.get(i);
				for (int ii = 0; ii < size; ii++) {
					if (item.equals(items[ii])) {
						removeIndex(ii);
						size--;
						break;
					}
				}
			}
		}
		return size != startSize;
	}

	public T pop() {
		if (size == 0)
			throw new IllegalStateException("TArray is empty.");
		--size;
		T item = items[size];
		items[size] = null;
		return item;
	}

	public T peek() {
		if (size == 0)
			throw new IllegalStateException("TArray is empty.");
		return items[size - 1];
	}

	public T first() {
		if (size == 0)
			throw new IllegalStateException("TArray is empty.");
		return items[0];
	}

	public void clear() {
		T[] items = this.items;
		for (int i = 0, n = size; i < n; i++)
			items[i] = null;
		size = 0;
	}

	public boolean isEmpty() {
		return this.size == 0;
	}

	public T[] shrink() {
		if (items.length != size)
			resize(size);
		return items;
	}

	public T[] ensureCapacity(int additionalCapacity) {
		int sizeNeeded = size + additionalCapacity;
		if (sizeNeeded > items.length)
			resize(MathUtils.max(8, sizeNeeded));
		return items;
	}

	protected T[] resize(int newSize) {
		T[] items = this.items;
		T[] newItems = (T[]) new Object[newSize];
		System.arraycopy(items, 0, newItems, 0,
				MathUtils.min(size, newItems.length));
		this.items = newItems;
		return newItems;
	}

	public void reverse() {
		T[] items = this.items;
		for (int i = 0, lastIndex = size - 1, n = size / 2; i < n; i++) {
			int ii = lastIndex - i;
			T temp = items[i];
			items[i] = items[ii];
			items[ii] = temp;
		}
	}

	public void shuffle() {
		T[] items = this.items;
		for (int i = size - 1; i >= 0; i--) {
			int ii = MathUtils.random(i);
			T temp = items[i];
			items[i] = items[ii];
			items[ii] = temp;
		}
	}

	public void truncate(int newSize) {
		if (size <= newSize)
			return;
		for (int i = newSize; i < size; i++)
			items[i] = null;
		size = newSize;
	}

	public Object last() {
		return items[size < 1 ? 0 : size - 1];
	}

	public T random() {
		if (size == 0)
			return null;
		return items[MathUtils.random(0, size - 1)];
	}

	public T[] toArray() {
		Object[] result = new Object[size];
		System.arraycopy(items, 0, result, 0, size);
		return (T[]) result;
	}

	public T[] toArray(T[] a) {
		int length = this.size;
		if (a.length < size) {
			a = (T[]) new Object[length];
		}
		Object[] result = a;
		for (int i = 0; i < length; ++i) {
			result[i] = get(i);
		}
		if (a.length > length) {
			a[length] = null;
		}
		return a;
	}

	private ArrayIterable iterable;

	public Iterator<T> iterator() {
		if (iterable == null) {
			iterable = new ArrayIterable(this);
		}
		return iterable.iterator();
	}

	public boolean equals(Object object) {
		if (object == this)
			return true;
		if (!(object instanceof TArray))
			return false;
		TArray array = (TArray) object;
		int n = size;
		if (n != array.size)
			return false;
		Object[] items1 = this.items;
		Object[] items2 = array.items;
		for (int i = 0; i < n; i++) {
			Object o1 = items1[i];
			Object o2 = items2[i];
			if (!(o1 == null ? o2 == null : o1.equals(o2)))
				return false;
		}
		return true;
	}

	public String toString() {
		if (size == 0)
			return "[]";
		T[] items = this.items;
		StringBuilder buffer = new StringBuilder(32);
		buffer.append('[');
		buffer.append(items[0]);
		for (int i = 1; i < size; i++) {
			buffer.append(", ");
			buffer.append(items[i]);
		}
		buffer.append(']');
		return buffer.toString();
	}

	public String toString(String separator) {
		if (size == 0) {
			return "";
		}
		T[] items = this.items;
		StringBuilder buffer = new StringBuilder(32);
		buffer.append(items[0]);
		for (int i = 1; i < size; i++) {
			buffer.append(separator);
			buffer.append(items[i]);
		}
		return buffer.toString();
	}

	static public <T> TArray<T> with(T... array) {
		return new TArray(array);
	}

	static public class ArrayIterator<T> implements Iterator<T>, Iterable<T> {

		private final TArray<T> array;
		private final boolean allowRemove;
		int index;
		boolean valid = true;

		public ArrayIterator(TArray<T> array) {
			this(array, true);
		}

		public ArrayIterator(TArray<T> array, boolean allowRemove) {
			this.array = array;
			this.allowRemove = allowRemove;
		}

		public boolean hasNext() {
			if (!valid) {
				throw new RuntimeException("#iterator() cannot be used nested.");
			}
			return index < array.size;
		}

		public T next() {
			if (index >= array.size) {
				throw new NoSuchElementException(String.valueOf(index));
			}
			if (!valid) {
				throw new RuntimeException("#iterator() cannot be used nested.");
			}
			return array.items[index++];
		}

		@Override
		public void remove() {
			if (!allowRemove) {
				throw new RuntimeException("Remove not allowed.");
			}
			index--;
			array.removeIndex(index);
		}

		public void reset() {
			index = 0;
		}

		public Iterator<T> iterator() {
			return this;
		}
	}

	static public class ArrayIterable<T> implements Iterable<T> {

		private final TArray<T> array;
		private final boolean allowRemove;
		private ArrayIterator iterator1, iterator2;

		public ArrayIterable(TArray<T> array) {
			this(array, true);
		}

		public ArrayIterable(TArray<T> array, boolean allowRemove) {
			this.array = array;
			this.allowRemove = allowRemove;
		}

		public Iterator<T> iterator() {
			if (iterator1 == null) {
				iterator1 = new ArrayIterator(array, allowRemove);
				iterator2 = new ArrayIterator(array, allowRemove);
			}
			if (!iterator1.valid) {
				iterator1.index = 0;
				iterator1.valid = true;
				iterator2.valid = false;
				return iterator1;
			}
			iterator2.index = 0;
			iterator2.valid = true;
			iterator1.valid = false;
			return iterator2;
		}
	}
}

File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/utils/TArray.cs
using java.lang;
using java.util;
using java.util.function;

namespace loon.utils
{

    public class TArray<T> : Iterable<T>, IArray
    {


        public class ArrayIterable<N> : Iterable<N>
        {

            private readonly TArray<N> array;
            private readonly bool allowRemove;
            private ArrayIterator<N> iterator1, iterator2;

            public ArrayIterable(TArray<N> array) : this(array, true)
            {

            }

            public ArrayIterable(TArray<N> array, bool allowRemove)
            {
                this.array = array;
                this.allowRemove = allowRemove;
            }

            public void ForEach(Consumer consumer)
            {
                Iterable_Java<N>.ForEach(this, consumer);
            }

            public Iterator<N> Iterator()
            {
                if (iterator1 == null)
                {
                    iterator1 = new ArrayIterator<N>(array, allowRemove);
                    iterator2 = new ArrayIterator<N>(array, allowRemove);
                }
                if (!iterator1.valid)
                {
                    iterator1.index = 0;
                    iterator1.valid = true;
                    iterator2.valid = false;
                    return iterator1;
                }
                iterator2.index = 0;
                iterator2.valid = true;
                iterator1.valid = false;
                return iterator2;
            }
        }

        public class ArrayIterator<N> : LIterator<N>, Iterable<N>
        {

            private readonly TArray<N> arrays;
            private readonly bool allowRemove;
            internal int index;
            internal bool valid = true;

            public ArrayIterator(TArray<N> array) : this(array, true)
            {

            }

            public ArrayIterator(TArray<N> array, bool allowRemove)
            {
                this.arrays = array;
                this.allowRemove = allowRemove;
            }


            public bool HasNext()
            {
                if (!valid)
                {
                    throw new LSysException("iterator() cannot be used nested.");
                }
                return index < arrays.size;
            }


            public N Next()
            {
                if (index >= arrays.size)
                {
                    return default;
                }
                if (!valid)
                {
                    throw new LSysException("iterator() cannot be used nested.");
                }
                return arrays.items[index++];
            }

            public void Remove()
            {
                if (!allowRemove)
                {
                    throw new LSysException("Remove not allowed.");
                }
                index--;
                arrays.RemoveIndex(index);
            }

            public void Reset()
            {
                index = 0;
            }

            public void ForEach(Consumer consumer)
            {
                Iterable_Java<N>.ForEach(this, consumer);
            }

            public Iterator<N> Iterator()
            {
                return this;
            }
        }

        public static TArray<T> At(int capacity)
        {
            return new TArray<T>(capacity);
        }

        public static TArray<T> At(TArray<T> array)
        {
            return new TArray<T>(array);
        }

        public static TArray<T> At()
        {
            return At(0);
        }
        public static TArray<T> With(params T[] array)
        {
            return new TArray<T>(array);
        }

        public void ForEach(Consumer consumer)
        {
            Iterable_Java<T>.ForEach(this, consumer);
        }

        public T[] items;

        public int size;

        public bool ordered;

        public TArray() : this(true, CollectionUtils.INITIAL_CAPACITY)
        {

        }

        public TArray(int capacity) : this(true, capacity)
        {

        }

        public TArray(bool ordered, int capacity)
        {
            this.ordered = ordered;
            items = new T[capacity];
        }

        public TArray(TArray<T> array) : this(array.ordered, array.size)
        {

            size = array.size;
            JavaSystem.Arraycopy(array.items, 0, items, 0, size);
        }

        public TArray(params T[] array) : this(true, array, 0, array.Length)
        {

        }

        public TArray(bool ordered, T[] array, int start, int count) : this(ordered, count)
        {
            size = count;
            JavaSystem.Arraycopy(array, start, items, 0, size);
        }

        public TArray(Array<T> vals) : this()
        {

            for (; vals.HashNext();)
            {
                Add(vals.Next());
            }
            vals.StopNext();
        }


        public bool Add(T value)
        {
            T[] items = this.items;
            if (size == items.Length)
            {
                items = Resize(MathUtils.Max(8, (int)(size * 1.75f)));
            }
            items[size++] = value;
            return true;
        }

        public void AddAll(TArray<T> array)
        {
            AddAll(array, 0, array.size);
        }

        public void AddAll(TArray<T> array, int start, int count)
        {
            if (start + count > array.size)
            {
                throw new LSysException("start + count must be <= size: " + start + " + " + count + " <= " + array.size);
            }
            AddAll(array.items, start, count);
        }

        public void AddAll(params T[] array)
        {
            AddAll(array, 0, array.Length);
        }

        public void AddAll(T[] array, int start, int count)
        {
            T[] items = this.items;
            int sizeNeeded = size + count;
            if (sizeNeeded > items.Length)
            {
                items = Resize(MathUtils.Max(8, (int)(sizeNeeded * 1.75f)));
            }
            JavaSystem.Arraycopy(array, start, items, size, count);
            size += count;
        }

        public T Get(int index)
        {
            if (index >= size)
                throw new LSysException("index can't be >= size: " + index + " >= " + size);
            return items[index];
        }

        public void Set(int index, T value)
        {
            if (index >= size)
                throw new LSysException("index can't be >= size: " + index + " >= " + size);
            items[index] = value;
        }

        public void Insert(int index, T value)
        {
            if (index > size)
                throw new LSysException("index can't be > size: " + index + " > " + size);
            T[] items = this.items;
            if (size == items.Length)
                items = Resize(MathUtils.Max(8, (int)(size * 1.75f)));
            if (ordered)
                JavaSystem.Arraycopy(items, index, items, index + 1, size - index);
            else
                items[size] = items[index];
            size++;
            items[index] = value;
        }

        public void Swap(int first, int second)
        {
            if (first >= size)
                throw new LSysException("first can't be >= size: " + first + " >= " + size);
            if (second >= size)
                throw new LSysException("second can't be >= size: " + second + " >= " + size);
            T[] items = this.items;
            T firstValue = items[first];
            items[first] = items[second];
            items[second] = firstValue;
        }

        public bool Contains(T value)
        {
            return Contains(value, false);
        }

        public bool Contains(T value, bool identity)
        {
            T[] items = this.items;
            int i = size - 1;
            if (identity || value == null)
            {
                while (i >= 0)
                    if ((object)items[i--] == (object)value)
                        return true;
            }
            else
            {
                while (i >= 0)
                    if (value.Equals(items[i--]))
                        return true;
            }
            return false;
        }

        public int IndexOf(T value)
        {
            return IndexOf(value, false);
        }

        public int IndexOf(T value, bool identity)
        {
            T[] items = this.items;
            if (identity || value == null)
            {
                for (int i = 0, n = size; i < n; i++)
                    if ((object)items[i] == (object)value)
                        return i;
            }
            else
            {
                for (int i = 0, n = size; i < n; i++)
                    if (value.Equals(items[i]))
                        return i;
            }
            return -1;
        }

        public int LastIndexOf(T value, bool identity)
        {
            T[] items = this.items;
            if (identity || value == null)
            {
                for (int i = size - 1; i >= 0; i--)
                    if ((object)items[i] == (object)value)
                        return i;
            }
            else
            {
                for (int i = size - 1; i >= 0; i--)
                    if (value.Equals(items[i]))
                        return i;
            }
            return -1;
        }

        public bool RemoveValue(T value)
        {
            T[] items = this.items;
            for (int i = 0, n = size; i < n; i++)
            {
                if ((object)items[i] == (object)value || value.Equals(items[i]))
                {
                    RemoveIndex(i);
                    return true;
                }
            }
            return false;
        }

        public bool RemoveValue(T value, bool identity)
        {
            T[] items = this.items;
            if (identity || value == null)
            {
                for (int i = 0, n = size; i < n; i++)
                {
                    if ((object)items[i] == (object)value)
                    {
                        RemoveIndex(i);
                        return true;
                    }
                }
            }
            else
            {
                for (int i = 0, n = size; i < n; i++)
                {
                    if (value.Equals(items[i]))
                    {
                        RemoveIndex(i);
                        return true;
                    }
                }
            }
            return false;
        }

        public T RemoveIndex(int index)
        {
            if (index >= size)
                throw new LSysException("index can't be >= size: " + index + " >= " + size);
            T[] items = this.items;
            T value = (T)items[index];
            size--;
            if (ordered)
                JavaSystem.Arraycopy(items, index + 1, items, index, size - index);
            else
                items[index] = items[size];
            items[size] = default;
            return value;
        }

        public void RemoveRange(int start, int end)
        {
            if (end >= size)
                throw new LSysException("end can't be >= size: " + end + " >= " + size);
            if (start > end)
                throw new LSysException("start can't be > end: " + start + " > " + end);
            T[] items = this.items;
            int count = end - start + 1;
            if (ordered)
                JavaSystem.Arraycopy(items, start + count, items, start, size - (start + count));
            else
            {
                int lastIndex = this.size - 1;
                for (int i = 0; i < count; i++)
                    items[start + i] = items[lastIndex - i];
            }
            size -= count;
        }

        public bool Remove(T value)
        {
            return Remove(value, false);
        }

        public bool Remove(T value, bool identity)
        {
            T[] items = this.items;
            if (identity || value == null)
            {
                for (int i = 0; i < size; i++)
                {
                    if ((object)items[i] == (object)value)
                    {
                        RemoveIndex(i);
                        return true;
                    }
                }
            }
            else
            {
                for (int i = 0; i < size; i++)
                {
                    if (value.Equals(items[i]))
                    {
                        RemoveIndex(i);
                        return true;
                    }
                }
            }
            return false;
        }

        public bool RemoveAll(TArray<T> array)
        {
            return RemoveAll(array, false);
        }

        public bool RemoveAll(TArray<T> array, bool identity)
        {
            if (array.size == 0)
            {
                return true;
            }
            int size = this.size;
            int startSize = size;
            T[] items = this.items;
            if (identity)
            {
                for (int i = 0, n = array.size; i < n; i++)
                {
                    T item = array.Get(i);
                    for (int ii = 0; ii < size; ii++)
                    {
                        if ((object)item == (object)items[ii])
                        {
                            RemoveIndex(ii);
                            size--;
                            break;
                        }
                    }
                }
            }
            else
            {
                for (int i = 0, n = array.size; i < n; i++)
                {
                    T item = array.Get(i);
                    for (int ii = 0; ii < size; ii++)
                    {
                        if (item.Equals(items[ii]))
                        {
                            RemoveIndex(ii);
                            size--;
                            break;
                        }
                    }
                }
            }
            return size != startSize;
        }

        public TArray<T> Cpy()
        {
            return new TArray<T>(items);
        }

        public T Pop()
        {
            if (size == 0)
                throw new LSysException("TArray is empty.");
            --size;
            T item = items[size];
            items[size] = default;
            return item;
        }

        public T Peek()
        {
            if (size == 0)
                throw new LSysException("TArray is empty.");
            return items[size - 1];
        }

        public T First()
        {
            if (size == 0)
                throw new LSysException("TArray is empty.");
            return items[0];
        }


        public void Clear()
        {
            T[] items = this.items;
            for (int i = 0, n = size; i < n; i++)
                items[i] = default;
            size = 0;
        }


        public bool IsEmpty()
        {
            return this.size == 0;
        }

        public T[] Shrink()
        {
            if (items.Length != size)
                Resize(size);
            return items;
        }

        public T[] EnsureCapacity(int additionalCapacity)
        {
            int sizeNeeded = size + additionalCapacity;
            if (sizeNeeded > items.Length)
                Resize(MathUtils.Max(8, sizeNeeded));
            return items;
        }

        protected T[] Resize(int newSize)
        {
            T[] items = this.items;
            T[] newItems = new T[newSize];
            JavaSystem.Arraycopy(items, 0, newItems, 0, MathUtils.Min(size, newItems.Length));
            this.items = newItems;
            return newItems;
        }

        public TArray<T> Reverse()
        {
            T[] items = this.items;
            for (int i = 0, lastIndex = size - 1, n = size / 2; i < n; i++)
            {
                int ii = lastIndex - i;
                T temp = items[i];
                items[i] = items[ii];
                items[ii] = temp;
            }
            return this;
        }

        public TArray<T> Shuffle()
        {
            T[] items = this.items;
            for (int i = size - 1; i >= 0; i--)
            {
                int ii = MathUtils.Random(i);
                T temp = items[i];
                items[i] = items[ii];
                items[ii] = temp;
            }
            return this;
        }

        public TArray<T> Unshift(T o)
        {
            T[] items = this.items;
            int len = items.Length;
            T[] newItems = new T[len + 1];
            newItems[0] = o;
            JavaSystem.Arraycopy(items, 0, newItems, 1, items.Length);
            this.items = newItems;
            this.size++;
            return this;
        }

        public void Truncate(int newSize)
        {
            if (size <= newSize)
                return;
            for (int i = newSize; i < size; i++)
                items[i] = default;
            size = newSize;
        }

        public T Last()
        {
            return items[size < 1 ? 0 : size - 1];
        }

        public T RemoveFirst()
        {
            return RemoveIndex(0);
        }

        public T RemoveLast()
        {
            return RemoveIndex(size < 1 ? 0 : size - 1);
        }

        public T Random()
        {
            if (size == 0)
                return default;
            return items[MathUtils.Random(0, size - 1)];
        }

        public TArray<T> RandomArrays()
        {
            if (size == 0)
            {
                return new TArray<T>();
            }
            T v = default;
            TArray<T> newArrays = new TArray<T>(size);
            for (int i = 0; i < size; i++)
            {
                newArrays.Add(Get(i));
            }
            for (int i = 0; i < size; i++)
            {
                v = Random();
                for (int j = 0; j < i; j++)
                {
                    if ((object)newArrays.Get(j) == (object)v)
                    {
                        v = Random();
                        j = -1;
                    }

                }
                newArrays.Set(i, v);
            }
            return newArrays;
        }

        public object[] ToArray()
        {
            object[] result = new object[size];
            JavaSystem.Arraycopy(items, 0, result, 0, size);
            return result;
        }

        public T[] ToArray(T[] a)
        {
            int length = this.size;
            if (a.Length < size)
            {
                a = new T[length];
            }
            T[] result = a;
            for (int i = 0; i < length; ++i)
            {
                result[i] = Get(i);
            }
            if (a.Length > length)
            {
                a[length] = default;
            }
            return a;
        }

        public override bool Equals(object o)
        {
            if (o == this)
                return true;
            if (!(o is TArray<T>))
            {
                return false;
            }
            TArray<T> array = (TArray<T>)o;
            int n = size;
            if (n != array.size)
                return false;
            T[] items1 = this.items;
            T[] items2 = array.items;
            for (int i = 0; i < n; i++)
            {
                T o1 = items1[i];
                T o2 = items2[i];
                if (!(o1 == null ? o2 == null : o1.Equals(o2)))
                    return false;
            }
            return true;
        }

        public TArray<T> Concat(TArray<T> array)
        {
            TArray<T> all = new TArray<T>(this);
            all.AddAll(array);
            return all;
        }


        private ArrayIterable<T> iterable;
        public Iterator<T> Iterator()
        {
            if (iterable == null)
            {
                iterable = new ArrayIterable<T>(this);
            }
            return iterable.Iterator();
        }



        public bool RetainAll(TArray<T> array)
        {
            T[] elementData = this.items;
            int r = 0, w = 0;
            bool modified = false;
            try
            {
                for (; r < size; r++)
                    if (array.Contains(elementData[r]))
                    {
                        elementData[w++] = elementData[r];
                    }
            }
            finally
            {
                if (r != size)
                {
                    JavaSystem.Arraycopy(elementData, r, elementData, w, size - r);
                    w += size - r;
                }
                if (w != size)
                {
                    for (int i = w; i < size; i++)
                    {
                        elementData[i] = default;
                    }
                    size = w;
                    modified = true;
                }
            }
            return modified;
        }

        public override int GetHashCode()
        {
            if (!ordered)
            {
                return base.GetHashCode();
            }
            int hashCode = 1;
            for (int i = size - 1; i > -1; i--)
            {
                hashCode = 31 * hashCode + (items[i] == null ? 0 : items[i].GetHashCode());
            }
            return hashCode;
        }


        public override string ToString()
        {
            if (size == 0)
                return "[]";
            T[] items = this.items;
            StrBuilder buffer = new StrBuilder(32);
            buffer.Append('[');
            buffer.Append(items[0]);
            for (int i = 1; i < size; i++)
            {
                buffer.Append(", ");
                buffer.Append(items[i]);
            }
            buffer.Append(']');
            return buffer.ToString();
        }

        public string ToString(string separator)
        {
            if (size == 0)
            {
                return "";
            }
            T[] items = this.items;
            StrBuilder buffer = new StrBuilder(32);
            buffer.Append(items[0]);
            for (int i = 1; i < size; i++)
            {
                buffer.Append(separator);
                buffer.Append(items[i]);
            }
            return buffer.ToString();
        }

        public int Size()
        {
            return size;
        }
    }
}
